File: speech_to_text/php/real_time/stream_file.php
<?php

require "vendor/autoload.php";

use Ratchet\Client\WebSocket;
use Ratchet\Client\Connector;
use Ratchet\RFC6455\Messaging\Frame;
use React\EventLoop\Factory;
use React\Socket\ConnectionException;
use React\Stream\ReadableResourceStream;

// Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
$api_key = getenv("SONIOX_API_KEY");
$ws_uri = "wss://stt-rt.soniox.com/transcribe-websocket";
$file_to_transcribe = "coffee_shop.pcm_s16le";

$loop = Factory::create();

$connector = new Connector($loop);

// Connect to WebSocket API
echo "Opening WebSocket connection..." . PHP_EOL;
$connector($ws_uri)->then(
    function (WebSocket $ws) use ($api_key, $loop, $file_to_transcribe) {
        // Send start request
        $ws->send(
            json_encode([
                "api_key" => $api_key,
                "audio_format" => "pcm_s16le",
                "sample_rate" => 16000,
                "num_channels" => 1,
                "model" => "stt-rt-preview",
                "language_hints" => array("en", "es")
            ])
        );

        $audio_file = fopen($file_to_transcribe, "rb");
        if (!$audio_file) {
            die("Failed to open audio file." . PHP_EOL);
        }

        echo "Transcription started." . PHP_EOL;

        // Read and send audio data from file over WebSocket connection every 120 ms
        $timer = $loop->addPeriodicTimer(0.12, function ($timer) use (
            $loop,
            $ws,
            $audio_file
        ) {
            $data = fread($audio_file, 3840);

            if (feof($audio_file) || empty($data)) {
                // Send end of file

                $ws->send("");
                $loop->cancelTimer($timer);
                return;
            }

            $ws->send(new Frame($data, true, Frame::OP_BINARY));
        });

        $final_text = "";

        $ws->on("message", function ($msg) use (&$final_text) {
            // Receive text
            $res = json_decode($msg, true);

            if (isset($res["error_code"])) {
                echo PHP_EOL . "Error: " . $res["error_code"] . " " . $res["error_message"] . PHP_EOL;
                return;
            }

            $non_final_text = "";

            foreach ($res["tokens"] as $token) {
                if (!empty($token["text"])) {
                    if ($token["is_final"]) {
                        $final_text = $final_text . $token["text"];
                    } else {
                        $non_final_text = $non_final_text . $token["text"];
                    }
                }
            }

            echo "\x1b[2J\x1b[H" . // clear the screen, move to top-left corner
                $final_text . // write final text
                "\x1b[34m" . // change text color to blue
                $non_final_text . // write non-final text
                "\x1b[39m"; // reset text color

            if (array_key_exists("finished", $res) && $res["finished"]) {
                echo PHP_EOL . "Transcription done." . PHP_EOL;
            }
        });

        $ws->on("close", function ($code = null, $reason = null) use (
            $loop,
            $timer
        ) {
            $loop->cancelTimer($timer);

            if ($reason != "") {
                echo "Close error: $code - $reason" . PHP_EOL;
            }
        });
    },
    function (ConnectionException $e) {
        echo "Could not connect: {$e->getMessage()}" . PHP_EOL;
    }
);

$loop->run();

?>

File: speech_to_text/java/real_time/stream_file/src/main/java/com/soniox/example/realtime/streamfile/StreamFile.java
package com.soniox.example.realtime.streamfile;

import java.io.*;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.net.URI;
import java.nio.*;
import java.util.concurrent.*;

import org.json.JSONObject;

public class StreamFile {
    public static void main(String[] args) {
        // Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
        String apiKey = System.getenv("SONIOX_API_KEY");
        URI websocketUrl = URI.create("wss://stt-rt.soniox.com/transcribe-websocket");
        String fileToTranscribe = "../../coffee_shop.pcm_s16le";

        ExecutorService executor = Executors.newCachedThreadPool();
        HttpClient client = HttpClient.newBuilder().executor(executor).build();

        CountDownLatch latch = new CountDownLatch(1);

        // Connect to WebSocket API
        System.out.println("Opening WebSocket connection...");

        WebSocket webSocket = client.newWebSocketBuilder().buildAsync(websocketUrl, new WebSocket.Listener() {
            String finalText = "";
            String nonFinalText = "";

            @Override
            public void onOpen(WebSocket webSocket) {
                // Send start request
                JSONObject startRequest = new JSONObject();
                startRequest.put("api_key", apiKey);
                startRequest.put("audio_format", "pcm_s16le");
                startRequest.put("sample_rate", 16000);
                startRequest.put("num_channels", 1);
                startRequest.put("model", "stt-rt-preview-v2");
                startRequest.put("language_hints", new String[]{"en", "es"});

                webSocket.sendText(startRequest.toString(), true);

                // Start send audio thread
                Thread sendAudioThread = new Thread(() -> {
                    // Send audio data
                    try (InputStream inputStream = new FileInputStream(fileToTranscribe)) {
                        System.out.println("Transcription started.");

                        try {
                            byte[] buffer = new byte[3840];
                            int bytesRead;

                            while ((bytesRead = inputStream.read(buffer)) != -1) {
                                webSocket.sendBinary(ByteBuffer.wrap(buffer, 0, bytesRead), true).join();
                                // Sleep for 120 ms
                                TimeUnit.MILLISECONDS.sleep(120);
                            }

                            // Send send of file
                            webSocket.sendText("", true).join();
                        } catch (IOException | InterruptedException | CompletionException e) {}
                    } catch (IOException e) {
                        e.printStackTrace();
                        System.exit(1);
                    }
                });
                executor.submit(sendAudioThread);

                WebSocket.Listener.super.onOpen(webSocket);
            }

            @Override
            public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
                // Receive text
                JSONObject response = new JSONObject(data.toString());

                // Handle any possible transcription error
                if (response.opt("error_code") != null)
                {
                    System.out.println("\nError: " + response.getInt("error_code") + ": " + response.getString("error_message"));
                    webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "");
                } else {
                    // Process tokens
                    nonFinalText = "";

                    response.getJSONArray("tokens").forEach(t -> {
                        JSONObject token = (JSONObject) t;
                        if (token.getString("text") != null) {
                            if (token.getBoolean("is_final")) {
                                finalText += token.getString("text");
                            } else {
                                nonFinalText += token.getString("text");
                            }
                        }
                    });

                    System.out.println(
                        "\033[2J\033[H" + // clear the screen, move to top-left corner
                        finalText + // write final text
                        "\033[34m" + // change text color to blue
                        nonFinalText + // write non-final text
                        "\033[39m" // reset text color
                    );

                    if (response.opt("finished") != null && response.getBoolean("finished")) {
                        // Close the WebSocket
                        System.out.println("\nTranscription done.");
                        webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "");
                    }
                }

                return WebSocket.Listener.super.onText(webSocket, data, last);
            }

            @Override
            public CompletionStage<?> onClose(WebSocket webSocket, int statusCode, String reason) {
                System.out.println("");

                if (!reason.equals("")) {
                    System.out.println("Close error: " + reason);
                }

                latch.countDown();

                return WebSocket.Listener.super.onClose(webSocket, statusCode, reason);
            }

            @Override
            public void onError(WebSocket webSocket, Throwable error) {
                System.out.println("");

                error.printStackTrace();

                latch.countDown();
            }
        }).join();

        try {
            latch.await();
        } catch (InterruptedException e) {}

        executor.shutdownNow();
    }
}
